Add Hadoop Converter Job and task

drcrallen opened the pull request:
* Allows submitting conversion tasks through Hadoop
* Allows submitting conversion tasks through the indexing service for Hadoop
* Adds unit tests for the hadoop conversion task. 
* Small changes to SQLMetadataSegmentManager to facilitate better unit tests
* Fixes https://github.com/druid-io/druid/issues/1363

The following should merge first:
https://github.com/druid-io/druid/pull/1367 (done)
https://github.com/druid-io/druid/pull/1366 (done)
https://github.com/druid-io/druid/pull/1428 (done)

Then I'll rebase this one.

xvrl commented on server/src/main/java/io/druid/metadata/storage/derby/DerbyEmbeddedMetadataStorageDruidModule.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Most useful for unit testing when `org.apache.derby.jdbc.EmbeddedDriver` is needed
+ */
+public class DerbyEmbeddedMetadataStorageDruidModule extends SQLMetadataStorageDruidModule
```
I'm not sure I understand why we need a new module here?

drcrallen replied:
`DerbyEmbeddedConnector` uses `org.apache.derby.jdbc.EmbeddedDriver` instead of the client driver. Another alternative is to have a setting that allows the configuration of the driver class.

drcrallen replied:
I managed to eliminate it and add a new test `@Rule` for anything that needs a metadata connector

cheddar commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1166 @@
+          log.info("Committing new segment");
+          taskContext.progress();
+          // Usually the indexing service handles its own commits, but standalone hadoop does not
+          if (shouldCommitMetadata(taskContext.getConfiguration())) {
```
I'm fairly certain this will attempt to write metadata as part of the reducer, which is actually not good.  A reducer can succeed and the total job fail, in which case we don't want the segment in the DB at all (it will be loaded by a historical node and just sit there, not able to do anything because its partitions weren't completed).  The segment metadata should be written all together in a single transaction to the DB once we know that the job has completed successfully.

drcrallen replied:
This is part of the task commit. At this point things are guaranteed to have been persisted to deep storage correctly, and the segment can be serviced independently. If it is desired to have the entire job be atomic, and not allow a completed segment to be serviced when others may not have been, then that is another concern.

Is there a reason why the job needs to be all or nothing?

cheddar replied:
> in which case we don't want the segment in the DB at all (it will be loaded by a historical node and just sit there, not able to do anything because its partitions weren't completed)

^ that's the reason... if you convert one partition and the others fail, for example, you now have zombie segments.

drcrallen replied:
I must be missing a critical detail, is there a check in the coordinator that prevents a segment from loading until all the partitions are accounted for?

drcrallen replied:
This was moved to the indexer

cheddar commented:
If I'm understanding this correctly, this looks like a way to run a conversion task over Hadoop MR.  That is, logically, it is executing tasks using Hadoop MR (and ultimately YARN) as the task manager instead of using whatever the Indexing Service is using.

I'm not against this, but I also kinda wonder if we shouldn't make a generic "run tasks on hadoop" job and then have the conversion task be a part of that?

drcrallen commented:
@cheddar : Yes, that is a longer goal.  Some of the aspects of this PR will be combined with the hadoop task once this has proven stable. A future goal would be to make a better task container that can run on the indexing service, yarn, or mesos.

drcrallen commented:
I had discussed with @xvrl briefly about what to do regarding the metadata update.

I'm in agreement with his point of view that only allowing the task to work as an indexing service task (and *NOT* as a standalone hadoop job) would greatly simply things overall.

@cheddar : is there any objection to simply having this as ONLY an indexing task?

xvrl commented:
The existing conversion tasks is also not standalone, so I don't see why this one would need to be.

It would also greatly simplify this PR to remove all metadata rated stuff, since that's unrelated to the task itself if we remove the standalone option.

drcrallen commented:
@cheddar / @xvrl : I removed the metadata updating in the job and left it as ONLY an indexing task.

I also fixed the last known bug (https://github.com/druid-io/druid/issues/1363) that I have encountered in local tests.

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1120 @@
+    return writtenBytes;
+  }
+
+  private static void setupClassPath(Job job, JobConf jobConf)
```
can we reuse code from JobHelper here ?

drcrallen replied:
No, not without modification. JobHelper relies exclusively on `HadoopDruidIndexerConfig` and not interfaces to something more abstract.

I tried not to touch any existing hadoop codepaths until this is stable because hadoop is incredibly picky.

xvrl replied:
@drcrallen it should be easy to reuse the JobHelper code for this. We can simply change the `JobHelper.setupClassPath` method to pass the workingPath as opposed to the entire config.

drcrallen replied:
I'll be happy to set it, but tried to touch the existing hadoop stuff as little as possible.

If this PR is generally agreed upon and it proves pretty stable in our tests then I'll be happy to migrate existing hadoop stuff over to this PR's "framework" of doing things.

xvrl replied:
I would rather use the existing code unless there is a good reason to rewrite things. Something little changes can make a big differences and for things that can be re-used I would prefer we take the old code unless there is a good reason to rewrite it, or because it would be difficult to refactor.

drcrallen replied:
I'm just going to refactor it as part of this PR.

drcrallen replied:
Splitting into its own PR

drcrallen replied:
Fixed having some more common stuff

xvrl commented on common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java:
```diff
@@ -41,6 +41,25 @@
   @JsonProperty("password")
   private PasswordProvider passwordProvider;
 
+  public MetadataStorageConnectorConfig(){
```
this class can probably removed, since we decided to make this a simple indexing task

drcrallen replied:
The class is not new, just the ability to specify the fields in tests.

drcrallen commented:
Travis failed due to https://github.com/druid-io/druid/issues/1393 restarting

drcrallen commented:
@cheddar / @xvrl This has been sitting here for 2 days and is blocking stuff on my side. Please either comment or merge.

xvrl commented on common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java:
```diff
@@ -41,6 +41,25 @@
   private PasswordProvider passwordProvider;
 
+  public MetadataStorageConnectorConfig(){
+    // NOOP
```
can we just call this(null, null, null, ...) to make it clear the new constructor is the main one?

xvrl commented on indexing-hadoop/src/test/java/io/druid/indexer/updater/HadoopConverterJobTest.java:
```diff
@@ -0,0 +1,421 @@
+                    "type", "static",
+                    "paths", tmpInputFile.getAbsolutePath()
+                ),
+                metadataStorageUpdaterJobSpc,
```
`MetadataStorageUpdaterJobSpec` is not required for hadoop jobs

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/ConvertSegmentTask.java:
```diff
@@ -192,46 +195,68 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
+        .keep(
+            new Function<DataSegment, Task>()
+            {
+              @Nullable
```
never null

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/DetermineHashedPartitionsJob.java:
```diff
@@ -102,7 +102,7 @@ public boolean run()
         groupByJob.setNumReduceTasks(config.getSegmentGranularIntervals().get().size());
       }
-      JobHelper.setupClasspath(config, groupByJob);
+      JobHelper.setupClasspath(JobHelper.distributedClassPath(config.getWorkingPath()), groupByJob);
```
do we need those classpath changes, now that we abstracted the hadoop task stuff?

drcrallen replied:
Here, yes. The abstractions were not perfect because some functions (like this JobHelper still took specific configs instead of specific distributed classpaths.

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -47,9 +57,76 @@
+   * @return An iterable of the interesting files (or empty iterable)
+   * @throws IOException
+   */
+  private static Iterable<File> findClassFiles(final File dir) throws IOException
```
this function doesn't appear to be used anywhere

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -47,9 +57,76 @@
+   * @param file The file of interest
+   * @return true if the file should be included, false otherwise
+   */
+  private static boolean interestingFile(final File file)
```
this function is also unneeded if findClassFiles is removed.

drcrallen replied:
Fixed

xvrl commented on server/src/test/java/io/druid/metadata/TestDerbyDB.java:
```diff
@@ -0,0 +1,156 @@
+  {
+    port = SocketUtil.findOpenPort(1621);
+    tmpPath = Files.createTempFile("derby", ".db");
+    tmpPath.toFile().delete();
```
why not just use the existing `TestDerbyConnector`? It uses Derby in-memory, which would remove the need for finding ports, checking connections, and cleaning up files.

drcrallen replied:
At the time I couldn't get the test connector to function in the context I needed in HadoopConverterJobTest

xvrl replied:
ok, in that case we can just use `new DBI("jdbc:derby:memory:druidTest;create=true")` or extract it into a static method in `TestDerbyConnector` so we can reuse it here.

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/HadoopConverterTask.java:
```diff
@@ -0,0 +1,408 @@
+  }
+
+  @JsonCreator
+  private static HadoopConverterTask createFromJson(
```
can we get some serde tests for this?

drcrallen replied:
Fixed

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/HadoopConverterTask.java:
```diff
@@ -0,0 +1,408 @@
+      @JsonProperty("classpathPrefix") String classpathPrefix
+  )
+  {
+    final boolean isForce = force == null ? false : force;
```
if we default to false we don't have to use Boolean

drcrallen replied:
fixed

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/HadoopConverterTask.java:
```diff
@@ -0,0 +1,408 @@
+            classpathPrefix
+        );
+      } else {
+        return create(
```
any specific reason we need a separate create method here?

drcrallen replied:
legacy : fixed

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/HadoopConverterTask.java:
```diff
@@ -0,0 +1,408 @@
+    return joinId(TYPE, dataSource, interval.getStart(), interval.getEnd(), new DateTime());
+  }
+
+  protected HadoopConverterTask(
```
can we do the json serde on the constructor and have a single entry point for task creation?

drcrallen replied:
This is a legacy from the ConvertTask but it also works around the oddities of having null segments which the original convert task supported. I'll give it a once over and see if I can simplify it.

drcrallen replied:
fixed

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+      }
+      try {
+        final Path inPath = new Path(getURIFromSegment(segment));
+        final FileSystem fs = inPath.getFileSystem(context.getConfiguration());
```
We probably would want to reuse the logic in `IndexGeneratorJob.serializeOutIndex` here, instead of reinventing it.

drcrallen replied:
The `serializeOutIndex` job for the `IndexGeneratorJob` is not pluggable at all. It is deeply interweaved with the IndexGeneratorJob way of doing things including the structure of its config.

It would be worth making that codepath more modular in general, but that feels like a pretty big scope creep for this PR unless it really is required.

xvrl replied:
I didn't mean to reuse all of `serializeOutIndex`, just extract some of the pieces into methods that we can reuse here.
For instance there is a long comments as to why we specifically don't to `instanceof` checks for filesystems in there, and I think we should reuse that part instead of doing it differently here.

drcrallen replied:
I made https://github.com/druid-io/druid/pull/1428 to handle this ask

drcrallen replied:
Done

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        } else {
+          fs.copyToLocalFile(false, inPath, new Path(tmpDownloadFile.toURI()));
+        }
+        // This is to bypass Guava dependencies in CompressionUtils.unzip
```
Maybe extract this into a `CompressionUtils.unzipNoGuava` until that's resolved?

drcrallen replied:
It feels weird to have methods that work around library problems depending on the library version used (rather than having a singular entry point).

It would look cleaner if the CompressionUtils auto-figured out if it could run regardless of guava version. But that reflection magic is completely lacking.

What would be NICER is if hadoop didn't require crappy old versions of guava.

xvrl replied:
that's fine, but we should at least extract it as a method, and put a comment as to why we can't use CompressionUtils, so that when that's resolved we can simply drop in the replacement.

drcrallen replied:
That I can do.

drcrallen replied:
Fixed

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+
+      // Make the zip file
+      final DataSegment finalSegment;
+      final File tmpZipFile = File.createTempFile("zipForUpload", ".zip");
```
we can probably reuse the code from `IndexGeneratorJob.serializeOutIndex` here to create and upload the zip file.
One nice property of that code is that it avoids the extra temporary file and writes the zip stream directly to the output target.

drcrallen replied:
It also cannot retry.

xvrl replied:
I'm more worried about having two different logics to deal with different file systems than the retry part.
We can easily add the retry `IndexGeneratorJob.serializeOutIndex` in a separate PR once we've extracted that functionality.

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/ConvertSegmentTask.java:
```diff
@@ -192,46 +193,67 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
+      final boolean validate
+  )
+  {
+    return FunctionalIterable
```
Iterables.transform ?

drcrallen replied:
fixed

nishantmonu51 commented on indexing-hadoop/pom.xml:
```diff
@@ -104,6 +104,58 @@
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-mapreduce-client-jobclient</artifactId>
+            <version>2.3.0</version>
```
probably extract the version as  a common property hadoop.version ?

fjy replied:
+1

drcrallen replied:
fixed. Added hadoop.compile.version

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+          )
+      );
+    } else {
+      final Set<String> segmentNames = Sets.newHashSet(
```
rename variable to dataSourceNames or dataSources ?

drcrallen replied:
fixed

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+
+
+  private static synchronized Injector getInjector(final Configuration configuration) throws IOException
+  {
```
injector doesn't seems to be used anywhere, can we just get rid of it ?

drcrallen replied:
fixed

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        cleanup(job);
+        return returnList;
+      } else {
+        throw new ISE(
```
should we cleanup workingDirectory before throwing exception?

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+            log.error("Error in task [%s] : %s", report.getTaskId(), Arrays.toString(report.getDiagnostics()));
+          }
+        }
+        cleanup(job);
```
I generally prefer to do cleanups in finally block, to make sure every code branch does the cleanup properly

fjy replied:
+1

drcrallen replied:
Here it only cleans up on failure. on success it leaves the data for the calling task to clean up

drcrallen replied:
oh wait, IR crazy. let me fix that

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+              taskContext.progress();
+            }
+          };
+          final HadoopDruidConverterConfig config = converterConfigFromConfiguration(taskContext.getConfiguration());
```
config is unused variable.

drcrallen replied:
fixed

xvrl commented on indexing-hadoop/pom.xml:
```diff
@@ -121,6 +173,16 @@
+                                <filter>
+                                    <artifact>*:*</artifact>
+                                    <excludes>
+                                        <exclude>META-INF/*.SF</exclude>
```
can we add a comment as to why we need this?

drcrallen replied:
It was weirding out when I was using hadoop self contained. But that should be in a different PR. Removing

nishantmonu51 commented on indexing-service/src/main/java/io/druid/indexing/common/task/HadoopConverterTask.java:
```diff
@@ -0,0 +1,276 @@
+    @Override
+    public boolean isReady(TaskActionClient taskActionClient) throws Exception
+    {
+      return false;
```
should this return true instead ?

nishantmonu51 replied:
or probably extend AbstractFixedIntervalTask ?

drcrallen replied:
The problem is that HadoopTask cannot extend AbstracFixedIntervalTask because HadoopIndexerTask has multiple intervals definable in the GranularitySpec.

drcrallen replied:
Fixed to return True.

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+    }
+    final List<DataSegment> segments = converterConfig.getSegments();
+    if (segments.isEmpty()) {
+      throw new IllegalArgumentException(
```
minor thing, but you can use IAE here. Make string substitution easier

drcrallen replied:
fixed

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        final TaskReport[] reports = job.getTaskReports(TaskType.MAP);
+        if (reports != null) {
+          for (final TaskReport report : reports) {
+            log.error("Error in task [%s] : %s", report.getTaskId(), Arrays.toString(report.getDiagnostics()));
```
is error enough? should this be emitted?

drcrallen replied:
HadoopIndexTask simply bubbles up success or failure, and doesn't even log.

himanshug replied:
I think logging the job url and success/failure should be good enough. failure would need more investigation and user could go to MR job logs to see all the details and diagnostics.

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        writtenBytes = job.getCounters().findCounter(COUNTER_GROUP, COUNTER_WRITTEN).getValue();
+      }
+      catch (IOException ex) {
+        log.error(ex, "Could not fetch counters");
```
should this be emitted?

drcrallen replied:
Probably not, if a JobHistory server is temporarily unavailable, it can encounter this error. Having each task emit alerts is probably overkill compared to a more general JobHistory checker.

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+                  try {
+                    if (!fs.exists(input)) {
+                      throw new ISE(
+                          String.format(
```
String.format is redundant

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+    );
+  }
+
+  public static Path getJobPath(JobID jobID, Path workingDirectory)
```
can we define static methods at the top of the file?

drcrallen replied:
Moved to top after static variables.

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        @Override
+        public void setupJob(JobContext jobContext) throws IOException
+        {
+
```
can we comment these as noop as well?

drcrallen replied:
fixed

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+              finalSegmentString,
+              DataSegment.class
+          );
+          log.info("Committing new segment");
```
can we also log what the segment is?

drcrallen replied:
fixed

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        @Override
+        public void abortTask(TaskAttemptContext taskContext) throws IOException
+        {
+          log.warn("Aborting task");
```
is warn enough?

drcrallen replied:
This is only logs for the OutputCommitter. Ideally you would cleanup any intermediate stuff here, but since our intermediate files are handled in the Mapper, there is nothing here to cleanup. This *should* be surrounded by other log messages indicating the reason for the abort and other informations outside the OutputCommitter

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+    {
+      final InputSplit split = context.getInputSplit();
+      if (!(split instanceof DataSegmentSplit)) {
+        throw new IOException(
```
IAE?

drcrallen replied:
fixed

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+      if (!outDir.mkdir() && (!outDir.exists() || !outDir.isDirectory())) {
+        throw new IOException(String.format("Could not create output directory [%s]", outDir));
+      }
+      IndexMaker.convert(
```
use IndexMerger

fjy replied:
we should look into removing IndexMaker. It was created to build v9 segments directly, but given the fact that incremental index is effectively a giant hashmap, a lot more work needs to be done before we can build v9 segments without conversion

drcrallen replied:
I have a separate commit for that

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException
+    {
+      final File tmpFile = Files.createTempDir();
```
how does this get cleaned up?

fjy replied:
nvm I see it

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+      final Map<String, Object> loadSpec = dataSegment.getLoadSpec();
+      final String type = loadSpec.get("type").toString();
+      final URI segmentLocURI;
+      if ("s3_zip".equals(type)) {
```
for anyone extending deep storage, no one will remember to add code here.

fjy replied:
Can the loadspec be in charge of creating these URIs?

drcrallen replied:
Currently that capacity resides in the DataPusher. 

I would like to see it reside in the DataPuller as well, because there is already a standard way to get a DataPuller from a LoadSpec. See the comment about URIDataPuller immediately prior.

drcrallen replied:
Also some of the functionality of DataPuller and DataPusher should eventually be combined into one FileObject type of thing

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        }
+      } else {
+        try {
+          throw new IAE(
```
there are other types of loadSpecs (e.g. azure)

drcrallen replied:
`io.druid.indexer.IndexGeneratorJob.IndexGeneratorReducer#serializeOutIndex`

xvrl replied:
@fjy as far as I know the hadoop index task has never supported anything that is not available as a filesystem in Hadoop. We should probably fix that and use DataSegmentPushers, but do it in the way that we can leverage native hadoop FS when it's faster than using the generic DataSegmentPusher way, as we currently are.

Doing so will also break backwards compatibility, since we currently rely on credentials in the hadoop config for the hadoop indexing, but rely on credentials in the Druid config for the DataSegmentPushers.

I'll file an issue for that

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+    ) throws IOException, InterruptedException
+    {
+      final InputSplit split = context.getInputSplit();
+      if (!(split instanceof DataSegmentSplit)) {
```
I hate instance of checks. Can we have this logic be inside the InputSplit and anything that is not a DataSegmentSplit throws an UOE?

drcrallen replied:
InputSplit is a Hadoop class

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,1021 @@
+        @Override
+        public String getCurrentValue() throws IOException, InterruptedException
+        {
+          readAnything = true;
```
What is the point of this class?

drcrallen replied:
Required for `InputFormat<String, String>` which is required to get the proper `InputSplit` on the segments to have each segment be 1 Mapper

fjy commented on indexing-service/src/main/java/io/druid/indexing/common/task/Task.java:
```diff
@@ -47,6 +47,8 @@
     @JsonSubTypes.Type(name = "index", value = IndexTask.class),
     @JsonSubTypes.Type(name = "index_hadoop", value = HadoopIndexTask.class),
+    @JsonSubTypes.Type(name = "hadoop_convert_segment", value = HadoopConverterTask.class),
+    @JsonSubTypes.Type(name = "hadoop_convert_segment_sub", value = HadoopConverterTask.ConverterSubTask.class),
```
i really think we shouldn't have sub tasks anymore

xvrl replied:
@fjy we should mark them as deprecated and explain what to use instead in the javadocs then. Would you mind submitting a PR for that so we can deprecate them in 0.8?

drcrallen replied:
Ok, I'm not deprecating them as part of this PR because that deserves its own discussion. And since this shares code with the Indexer Converter, it has to maintain compatibility until that one is also changed.

fjy replied:
it is fine to have them for now, at some point we should have a longer discussion about subtasks

xvrl replied:
@fjy do you mind filing an issue?

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,995 @@
+    }
+  }
+
+  public static class DataSegmentSplit extends InputSplit
```
it is more conventional to implement "org.apache.hadoop.io.Writable" as well so that you don't have to create and setup serde separately. I think that will reduce some code(DataSegmentSplitSerializer) and setupSerializers(..) method.

drcrallen replied:
fixed

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,995 @@
+    }
+  }
+
+  private static void setupSerializers(JobConf jobConf)
```
not needed if you implement Writable in DataSegmentSplit

drcrallen replied:
fixed

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,995 @@
+      );
+    }
+    converterConfigIntoConfiguration(converterConfig, segments, jobConf);
+    jobConf.setNumMapTasks(segments.size());
```
you shouldn't set this. this will automatically be the number of input splits

drcrallen replied:
Which should be equal to this, but I'll get rid of it anyways.

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,995 @@
+    converterConfigIntoConfiguration(converterConfig, segments, jobConf);
+    jobConf.setNumMapTasks(segments.size());
+    jobConf.setNumReduceTasks(0);
+    jobConf.setWorkingDirectory(new Path(converterConfig.getDistributedSuccessCache()));
```
what is the advantage of setting this?

drcrallen replied:
Because working directory is a first class configuration object, and the distributed success cache is used as a working directory.

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java:
```diff
@@ -0,0 +1,995 @@
+      throw runtimeException;
+    }
+    catch (RuntimeException ex) {
+      runtimeException = ex;
```
why not just use "throwable" to keep reference of exception as well and variable runtimeException is probably redundant.

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -425,4 +440,108 @@ public static Path prependFSIfNullScheme(FileSystem fs, Path path)
   }
+
+  // TODO: Replace this whenever hadoop gets their act together and stops breaking with more recent versions of Guava
+  public static long unzipNoGuava(
```
lol
+1 on this

fjy replied:
how much of this method is copypasta from guava?

drcrallen replied:
Zero, it is actually because `com.metamx.common.CompressionUtils` uses byte source and byte sink in the "useful" methods. Also `com.google.common.io.ByteStreams#copy(java.io.InputStream, java.io.OutputStream)` is used quite a bit in there.

xvrl commented on indexing-hadoop/pom.xml:
```diff
@@ -104,6 +104,58 @@
+        <dependency>
+            <groupId>org.apache.derby</groupId>
+            <artifactId>derby</artifactId>
+            <version>10.11.1.1</version>
```
derbynet in the parent pom already depends on derby, so we can just use derbynet and remove the version here.

drcrallen replied:
was only needed when the derby server rule was present. since it has vanished this can to. Will fix.

drcrallen replied:
fixed

xvrl commented:
+1

cheddar commented:
Conversations directly with FJ and others who have commented on this PR show that everyone is generally ok with this, so I went ahead and merged.
